clients/py/sapphirepy/sapphire.py

```diff
@@ -63,79 +87,152 @@ def _encrypt_tx_params(pk:CalldataPublicKey, params:tuple[TxParams]):
-                #  'core: invalid call format: Tag verification failed'
-                #  'core: invalid call format: epoch in the future'
-                # We can only do something meaningful with the first!
-                result = cast(RPCResponse, make_request(method, params))
-                if result.get('error', None) is not None:
-                    error = result['error']
-                    if not isinstance(error, str) and error['code'] == -32000:
-                        if error['message'] == 'core: invalid call format: epoch too far in the past':
-                            # force the re-fetch, and encrypt with new key
-                            do_fetch = True
-                            pk = None
-                            continue
-
-            # Only eth_call is decrypted
-            if method == 'eth_call' and result.get('result', '0x') != '0x':
-                decrypted = c.decrypt(unhexlify(result['result'][2:]))
-                result['result'] = HexStr('0x' + hexlify(decrypted).decode('ascii'))
-
-            return result
-        return make_request(method, params)
-    return middleware
-
-def wrap(w3: Web3):
+    def sapphire_middleware(
+            make_request: Callable[[RPCEndpoint, Any], Any], w3: "Web3"
+    ) -> Callable[[RPCEndpoint, Any], RPCResponse]:
+        """
+        Transparently encrypt the calldata for:
+
+         - eth_estimateGas
+         - eth_sendTransaction
+         - eth_call
+
+        The calldata public key, which used to derive a shared secret with an
+        ephemeral key, is retrieved upon the first request. This key is rotated by
+        Sapphire every epoch, and only transactions encrypted with keys from the
+        last 5 epochs are considered valid.
+
+        Deployment transactions will not be encrypted, unless the global
+        ENCRYPT_DEPLOYS flag is set. Encrypting deployments will prevent contracts
+        from being verified.
+
+        Pre-signed transactions can't be encrypted if submitted via this instance.
+        """
+        manager = CalldataPublicKeyManager()
+
+        def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
+            if _should_intercept(method, params):
+                do_fetch = True
+                pk = manager.newest
+                while do_fetch:
+                    if not pk:
+                        # If no calldata public key exists, fetch one
+                        cdpk = cast(RPCResponse, make_request(RPCEndpoint('oasis_callDataPublicKey'), []))
+                        pk = cast(Optional[CalldataPublicKey], cdpk.get('result', None))
+                        if pk:
+                            manager.add(pk)
+                    if not pk:
+                        raise RuntimeError('Could not retrieve callDataPublicKey!')
+                    do_fetch = False
+
+                    c = _encrypt_tx_params(pk, params, w3,  account)
```

Comment:
This makes no distinction between transactions being submitted and `eth_call` . This will break transactions - need two code paths, one for transactions and one for `eth_call`.